perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AuthzResolverBlImpl.java

```diff
@@ -1922,4 +1986,164 @@ private static void prepareServiceRoles(PerunSession sess) {
+	 *
+	 * @param objects for which will be reliable objects fetched.
+	 * @return all reliable objects together with the objects from the input.
+	 */
+	private static List<PerunBean> fetchAllReliableObjects(List<PerunBean> objects) {
+		List<PerunBean> reliableObjects = new ArrayList<>();
+
+		for (PerunBean object: objects) {
+			if (!reliableObjects.contains(object)) reliableObjects.add(object);
+			try {
+				List<PerunBean> retrievedObjects = PerunBeanName.valueOf(object.getBeanName()).apply(object);
+				retrievedObjects.forEach(o -> {
+					if (!reliableObjects.contains(o)) reliableObjects.add(o);
+				});
+			} catch (IllegalArgumentException ex) {
+				//It means that this PerunBean has no reliable objects thus it is not defined in the PerunBeanName
+			}
+		}
+		return reliableObjects;
+	}
+
+	/**
+	 * Enum defines PerunBean's name and action. The action retrieves all reliable objects for the object with that name.
+	 */
+	private enum PerunBeanName implements Function<PerunBean, List<PerunBean>> {
+		Member((object) -> {
+			User user = new User(((Member) object).getUserId(), "", "", "", "", "");
+			Vo vo = new Vo(((Member) object).getVoId(), "", "");
+			return Arrays.asList(user,vo);
+		}),
+		Group((object) -> {
+			Vo vo = new Vo(((Group) object).getVoId(), "", "");
+			return Collections.singletonList(vo);
+		}),
+		Resource((object) -> {
+			Vo vo = new Vo(((Resource) object).getVoId(), "", "");
+			Facility facility = new Facility(((Resource) object).getFacilityId(), "");
+			return Arrays.asList(vo, facility);
+		}),
+		ResourceTag((object) -> {
+			Vo vo = new Vo(((ResourceTag) object).getVoId(), "", "");
+			return Collections.singletonList(vo);
+		}),
+		RichMember((object) -> {
+			User user = new User(((RichMember) object).getUserId(), "", "", "", "", "");
+			Vo vo = new Vo(((RichMember) object).getVoId(), "", "");
+			return Arrays.asList(user,vo);
+		}),
+		RichGroup((object) -> {
+			Vo vo = new Vo(((RichGroup) object).getVoId(), "", "");
+			return Collections.singletonList(vo);
+		}),
+		RichResource((object) -> {
+			Vo vo = new Vo(((RichResource) object).getVoId(), "", "");
+			Facility facility = new Facility(((RichResource) object).getFacilityId(), "");
+			return Arrays.asList(vo, facility);
+		});
+
+		private Function<PerunBean, List<PerunBean>> function;
+
+		PerunBeanName(final Function<PerunBean, List<PerunBean>> function) {
+			this.function = function;
+		}
+
+		@Override
+		public List<PerunBean> apply(PerunBean object) {
+			return function.apply(object);
+		}
+	}
+
+	/**
+	 * Fetch policy_roles from the JsonNode object.
+	 *
+	 * @param policy from which will be policy roles fetched.
+	 * @return all policy roles nodes for the given policy.
+	 */
+	private static List<JsonNode> fetchPolicyRoles(JsonNode policy) {
+		JsonNode policyRolesNode = policy.get("policy_roles");
+		return new ArrayList<>(new ObjectMapper().convertValue(policyRolesNode, new TypeReference<List<JsonNode>>() {}));
```

Comment:
``ObjectMapper`` should be thread safe and can be static (unless you use different config options, but it doesn't seem to be that way).

Reply:
Fixed.